cpp/src/parquet/encryption/properties_driven_crypto_factory.h

```diff
@@ -0,0 +1,223 @@
+          cache_lifetime_seconds_(kDefaultCacheLifetimeSeconds),
+          internal_key_material_(kDefaultInternalKeyMaterial),
+          uniform_encryption_(kDefaultUniformEncryption),
+          data_key_length_bits_(kDefaultDataKeyLengthBits) {}
+
+    /// List of columns to encrypt, with master key IDs (see HIVE-21848).
+    /// Format: "masterKeyID:colName,colName;masterKeyID:colName..."
+    /// Either
+    /// column_keys(const std::string&)
+    /// or
+    /// uniform_encryption()
+    /// must be called. If none are called, or if both are called, an exception will be
+    /// thrown.
+    Builder* column_keys(const std::string& column_keys);
+
+    /// encrypt footer and columns with the same encryption key
+    Builder* uniform_encryption();
+
+    /// Parquet encryption algorithm. Can be "AES_GCM_V1" (default), or "AES_GCM_CTR_V1".
+    Builder* encryption_algorithm(ParquetCipher::type algo);
+
+    /// Write files with plaintext footer
+    Builder* plaintext_footer(bool plaintext_footer);
+
+    /// Use double wrapping - where data encryption keys (DEKs) are encrypted with key
+    /// encryption keys (KEKs), which in turn are encrypted with master keys.
+    Builder* double_wrapping(bool double_wrapping);
+
+    /// Wrap keys locally - master keys are fetched from the KMS server and used to
+    /// encrypt other keys (DEKs or KEKs).
+    Builder* wrap_locally(bool wrap_locally);
+
+    /// Lifetime of cached entities (key encryption keys, local wrapping keys, KMS client
+    /// objects).
+    Builder* cache_lifetime_seconds(uint64_t cache_lifetime_seconds);
+
+    /// Store key material inside Parquet file footers; this mode doesn’t produce
+    /// additional files. By default, true. If set to false, key material is stored in
+    /// separate files in the same folder, which enables key rotation for immutable
+    /// Parquet files.
+    Builder* internal_key_material(bool internal_key_material);
+
+    /// Length of data encryption keys (DEKs), randomly generated by parquet key
+    /// management tools. Can be 128, 192 or 256 bits.
+    Builder* data_key_length_bits(int32_t data_key_length_bits);
+
+    std::shared_ptr<EncryptionConfiguration> build();
+
+   private:
+    std::string footer_key_;
+    std::string column_keys_;
+    ParquetCipher::type encryption_algorithm_;
+    bool plaintext_footer_;
+    bool double_wrapping_;
+    bool wrap_locally_;
+    uint64_t cache_lifetime_seconds_;
+    bool internal_key_material_;
+    bool uniform_encryption_;
+    int32_t data_key_length_bits_;
+  };
+
+  const std::string& footer_key() const { return footer_key_; }
+  const std::string& column_keys() const { return column_keys_; }
+  ParquetCipher::type encryption_algorithm() const { return encryption_algorithm_; }
+  bool plaintext_footer() const { return plaintext_footer_; }
+  bool double_wrapping() const { return double_wrapping_; }
+  bool wrap_locally() const { return wrap_locally_; }
```

Comment:
As before (we need to remove this option. We won't be able to encrypt some of the files with local wrapping, and other files with remote wrapping.)